cmd/tke-installer/app/installer/application.go

```diff
@@ -0,0 +1,179 @@
+	"tkestack.io/tke/cmd/tke-installer/app/config"
+	"tkestack.io/tke/cmd/tke-installer/app/installer/constants"
+)
+
+const (
+	applicationInstallNamespaceDefault      = "default"
+	applicationInstallClusterDefault        = "global"
+	applicationInstallTenantIDDefault       = constants.DefaultTeantID
+	applicationInstallChartGroupNameDefault = constants.DefaultChartGroupName
+	applicationInstallDriverTypeDefault     = "HelmV3"
+	applicationInstallValueTypeDefault      = "yaml"
+)
+
+func (t *TKE) completePlatformApps() error {
+
+	if len(t.Config.PlatformApps) == 0 {
+		return nil
+	}
+
+	for _, platformApp := range t.Config.PlatformApps {
+		if !platformApp.Enable {
+			continue
+		}
+		err := t.completeChart(&platformApp.Chart)
+		if err != nil {
+			return fmt.Errorf("bad platform app config. %v, %v", platformApp.Name, err)
+		}
+	}
+
+	t.log.Infof("tke platform apps config completed.")
+	t.backup()
+	return nil
+}
+
+func (t *TKE) completeChart(chart *config.Chart) error {
+
+	_, err := chart.Values.YAML()
+	if err != nil {
+		return err
+	}
+
+	if chart.Name == "" {
+		return fmt.Errorf("chart name empty")
+	}
+	if chart.Version == "" {
+		return fmt.Errorf("chart value empty")
+	}
+	if chart.TenantID == "" {
+		chart.TenantID = applicationInstallTenantIDDefault
+	}
+	if chart.ChartGroupName == "" {
+		chart.ChartGroupName = applicationInstallChartGroupNameDefault
+	}
+	if chart.TargetCluster == "" {
+		chart.TargetCluster = applicationInstallClusterDefault
+	}
+	if chart.TargetNamespace == "" {
+		chart.TargetCluster = applicationInstallNamespaceDefault
+	}
+
+	return nil
+}
+
+func (t *TKE) installApplications(ctx context.Context) error {
+
+	if len(t.Config.PlatformApps) == 0 {
+		return nil
+	}
+
+	// TODO: (workaround ) client init will only be called in createCluster.
+	// If we SKIP createCluster step, all client calls will be panic
+	err := t.initDataForDeployTKE()
+	if err != nil {
+		return err
+	}
+
+	apps, err := t.applicationClient.Apps("").List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("list all applications failed %v", err)
+	}
+
+	for _, platformApp := range t.Config.PlatformApps {
+		if !platformApp.Enable {
+			continue
+		}
+		err := t.installApplication(ctx, platformApp, apps.Items)
+		if err != nil {
+			return fmt.Errorf("install application failed. %v, %v", platformApp.Name, err)
+		}
+		t.log.Infof("finish application installation %v", platformApp.Name)
+	}
+
+	return nil
+}
+
+func (t *TKE) installApplication(ctx context.Context, platformApp config.PlatformApp, installedApps []applicationv1.App) error {
+
+	chart := platformApp.Chart
+
+	var found bool
+	var duplicated bool
+	for _, installedApp := range installedApps {
+		// if there's an existed application with the same namespace+name, we consider it as already exists
+		if chart.Name == installedApp.Spec.Name && chart.TargetNamespace == installedApp.Namespace {
+			found = true
```

Comment:
found value is not necessary, should add log info and return here.

Reply:
removed